File: commands/src/main/java/kaap/veiko/debuggerforker/commands/commandsets/event/events/ThreadDeathEvent.java
package kaap.veiko.debuggerforker.commands.commandsets.event.events;

import com.google.auto.value.AutoValue;

import kaap.veiko.debuggerforker.types.DataReadException;
import kaap.veiko.debuggerforker.types.DataReader;
import kaap.veiko.debuggerforker.types.DataWriter;
import kaap.veiko.debuggerforker.types.jdwp.EventKind;
import kaap.veiko.debuggerforker.types.jdwp.ThreadId;

@AutoValue
public abstract class ThreadDeathEvent extends VirtualMachineEvent {
  public abstract int getRequestId();
  public abstract ThreadId getThread();

  public EventKind getEventKind() {
    return EventKind.THREAD_DEATH;
  }

  public static ThreadDeathEvent create(int requestId, ThreadId thread) {
    return new AutoValue_ThreadDeathEvent(requestId, thread);
  }

  public static ThreadDeathEvent read(DataReader reader) throws DataReadException {
    return create(
        reader.readInt(),
        ThreadId.read(reader)
    );
  }

  @Override
  public void write(DataWriter writer) {
    writer.writeType(getEventKind());
    writer.writeInt(getRequestId());
    writer.writeType(getThread());
  }
}

File: commands/src/main/kotlin/kaap/veiko/debuggerforker/commands/commandsets/event/events/ThreadDeathEvent.kt
package kaap.veiko.debuggerforker.commands.commandsets.event.events

import kaap.veiko.debuggerforker.types.DataReader
import kaap.veiko.debuggerforker.types.DataWriter
import kaap.veiko.debuggerforker.types.jdwp.EventKind
import kaap.veiko.debuggerforker.types.jdwp.ThreadId

data class ThreadDeathEvent(
        val requestId: Int,
        val thread: ThreadId
) : VirtualMachineEvent() {

    override fun write(writer: DataWriter?) = writer!!.run {
        writeType(EventKind.THREAD_DEATH)
        writeInt(requestId)
        writeType(thread)
    }

    companion object {
        @JvmStatic
        fun read(reader: DataReader) = ThreadDeathEvent(
                reader.readInt(),
                ThreadId.read(reader)
        )
    }
}